ixmp4/core/optimization/table.py

```diff
@@ -0,0 +1,121 @@
+from datetime import datetime
+from typing import Any, ClassVar, Iterable
+
+import pandas as pd
+
+from ixmp4.core.base import BaseFacade, BaseModelFacade
+from ixmp4.data.abstract import Docs as DocsModel
+from ixmp4.data.abstract import Run
+from ixmp4.data.abstract import Table as TableModel
+from ixmp4.data.abstract.optimization import Column
+
+
+class Table(BaseModelFacade):
+    _model: TableModel
+    NotFound: ClassVar = TableModel.NotFound
+    NotUnique: ClassVar = TableModel.NotUnique
+
+    @property
+    def id(self) -> int:
+        return self._model.id
+
+    @property
+    def name(self) -> str:
+        return self._model.name
+
+    @property
+    def run_id(self) -> int:
+        return self._model.run__id
+
+    @property
+    def data(self) -> dict[str, Any]:
+        return self._model.data
+
+    def add(self, data: dict[str, Any] | pd.DataFrame) -> None:
+        """Adds data to an existing Table."""
+        self.backend.optimization.tables.add_data(table_id=self._model.id, data=data)
+        self._model.data = self.backend.optimization.tables.get(
+            run_id=self._model.run__id, name=self._model.name
+        ).data
+
+    @property
+    def constrained_to_indexsets(self) -> list[str]:
+        return [column.indexset.name for column in self._model.columns]
+
+    @property
+    def columns(self) -> list[Column]:
+        return self._model.columns
+
+    @property
+    def created_at(self) -> datetime | None:
+        return self._model.created_at
+
+    @property
+    def created_by(self) -> str | None:
+        return self._model.created_by
+
+    @property
+    def docs(self):
+        try:
+            return self.backend.optimization.tables.docs.get(self.id).description
+        except DocsModel.NotFound:
+            return None
+
+    @docs.setter
+    def docs(self, description):
+        if description is None:
+            self.backend.optimization.tables.docs.delete(self.id)
+        else:
+            self.backend.optimization.tables.docs.set(self.id, description)
+
+    @docs.deleter
+    def docs(self):
+        try:
+            self.backend.optimization.tables.docs.delete(self.id)
+        # TODO: silently failing
+        except DocsModel.NotFound:
+            return None
+
+    def __str__(self) -> str:
+        return f"<Table {self.id} name={self.name}>"
+
+
+class TableRepository(BaseFacade):
+    _run: Run
+
+    def __init__(self, _run: Run, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self._run = _run
+
+    # TODO Here, constrained_to_indexsets is a list[str], while above it's list[int]
+    # We should make this consistent
+    def create(
+        self,
+        name: str,
+        constrained_to_indexsets: list[str],
+        column_names: list[str] | None = None,
```

Comment:
Shouldn't this be flipped around? column-names are required, constrained_to_indexsets is optional? And also changing the order would be more intuitive.

```suggestion
        column_names: list[str],
        constrained_to_indexsets: list[str] | None = None,
```

Reply:
To quote my reply from when you first brought this up:

> This is currently not possible because `column_names` has a default value while `constrained_to_indexsets` has not (I think this would result in a linting error, though ruff doesn't immediately flag it for me). The reason for this is that we take the column names from the names of the indexsets in `constrained_to_indexsets` per default and only allow overwriting them with `column_names`. We could change our approach to this, but it makes sense for me to offer the option to provide names only once if they correspond to `Indexset.name`s.

Please keep in mind that we are currently requiring every `Column` of a `Table` to be linked to an `IndexSet` for validation. Your suggestion here changes the name of the arguments, so we would require an argument called `column_names` and optionally take `constrained_to_indexsets` as well. We would then presumably try to get an `IndexSet` for each name in `column_names` and link it to the `Column`. If we don't find any `IndexSet` with that name, we would suddenly require `constrained_to_indexsets` to be there and provide existing names for every `Column` (unless we change that to be a mapping, which might be a good idea). 
In the current approach, we require `constrained_to_indexset` to always be there with existing `IndexSet.name`s, but this is a transparent requirement: it's just always required. Per default, we use the `IndexSet.name`s for the `Column` names, too, but you can still overwrite them if you prefer. 

I don't know, the current approach is more intuitive to me.

Reply:
Thanks for explaining it again, all good to proceed as is.